server/src/main/java/io/crate/types/CharacterType.java

```diff
@@ -214,11 +214,26 @@ public void addMappingOptions(Map<String, Object> mapping) {
 
     @Override
     public int compare(String val1, String val2) {
-        return val1.stripTrailing().compareTo(val2.stripTrailing());
+        // pads for values shorter than the length limit and also strips excess blank padding which should be ignored
+        // when compared.
+        return stripTrailingBlankPadding(padEnd(val1, lengthLimit, ' '))
+            .compareTo(stripTrailingBlankPadding(padEnd(val2, lengthLimit, ' ')));
     }
 
     @Override
     public StorageSupport<Object> storageSupport() {
         return storageSupport;
     }
+
+    private String stripTrailingBlankPadding(String s) {
+        int idx = s.length();
+        for (int i = idx - 1; i >= lengthLimit; i--) {
+            if (Character.isSpaceChar(s.charAt(i))) {
+                idx = i;
+            } else {
+                break;
+            }
+        }
+        return idx != s.length() ? s.substring(0, idx) : s;
```

Comment:
I'd simply return `s.substring(0, idx)` since the method checks and returns `this` if `idx` matches the length of the string.